File: go/data-structure/heap.go
package datastructure

import "math"

type Heap struct {
	arr  []int
	size int
}

func NewHeap() *Heap {
	return &Heap{arr: []int{}, size: 0}
}

func (h *Heap) getParentIndex(i int) int {
	return i / 2
}

func (h *Heap) getLeftChildIndex(i int) int {
	return 2 * i
}

func (h *Heap) getRightChildIndex(i int) int {
	return 2*i + 1
}

func (h *Heap) Size() int {
	return h.size
}

func (h *Heap) IsEmpty() bool {
	return h.size == 0
}

func (h *Heap) Heapfy(i int) {
	left := h.getLeftChildIndex(i)
	right := h.getRightChildIndex(i)
	largest := i

	if left < h.size && h.arr[left] > h.arr[i] {
		largest = left
	}

	if right < h.size && h.arr[right] > h.arr[largest] {
		largest = right
	}

	if i != largest {
		h.arr[i], h.arr[largest] = h.arr[largest], h.arr[i]
		h.Heapfy(largest)
	}
}

func (h *Heap) increaseKey(index int, newValue int) {
	h.arr[index] = newValue
	for index != 0 && h.arr[index] > h.arr[h.getParentIndex(index)] {
		h.arr[index], h.arr[h.getParentIndex(index)] = h.arr[h.getParentIndex(index)], h.arr[index]
		index = h.getParentIndex(index)
	}
}

func (h *Heap) Insert(i int) {
	h.arr = append(h.arr, i)
	h.size++
	currIndex := h.size

	if currIndex == 1 {
		return
	}

	for currIndex != 0 && h.arr[h.getParentIndex(currIndex)] < h.arr[currIndex] {
		h.arr[currIndex], h.arr[h.getParentIndex(currIndex)] = h.arr[h.getParentIndex(currIndex)], h.arr[currIndex]
		currIndex = h.getParentIndex(currIndex)
	}
}

func (h *Heap) Pop() int {
	if h.IsEmpty() {
		println("Heap is Empty")
		return math.MinInt
	}

	h.size--
	root := h.arr[0]
	h.arr[0] = h.arr[h.size-1]

	if !h.IsEmpty() {
		h.Heapfy(0)
	}

	return root
}

func (h *Heap) RemoveIndex(index int) {
	if index > h.size || index < 0 {
		println("Index Out Of Inbound")
		return
	}

	h.increaseKey(index, math.MaxInt)
	h.Pop()
}

File: cpp/DataStructure/Heap/Heap.cpp
#include <iostream>

class MaxHeap
{
    private:
        int* arr;
        int maxSize;
        int heapSize;

        void swap(int x, int y)
        {
            int temp = arr[x];
            arr[x] = arr[y];
            arr[y] = temp;
        }

    public:
        MaxHeap(int maxSize)
        {
            heapSize = 0;
            this->maxSize = maxSize;
            arr = new int[maxSize];
        }

        int parentIndex(int i)
        {
            return (i-1)/2;
        }

        int leftChildIndex(int i)
        {
            return 2*i+1;
        }

        int rightChildIndex(int i)
        {
            return 2*i+2;
        }

        int getMax()
        {
            return arr[0];
        }

        int size()
        {
            return heapSize;
        }

        void heapfy(int i)
        {
            int left = leftChildIndex(i);
            int right = rightChildIndex(i);
            int largest = i;

            if(left < heapSize && arr[left] > arr[i]) largest = left;
            if(right < heapSize && arr[right] > arr[largest]) largest = right;
            if (largest != i)
            {
                swap(i, largest);
                heapfy(largest);
            }
        }

        void increaseKey(int index, int newValue)
        {
            arr[index] = newValue;
            while (index != 0 && arr[index] > arr[parentIndex(index)])
            {
                swap(index, parentIndex(index));
                index = parentIndex(index);
            }
            
        }

        void insert(int x)
        {
            if (heapSize == maxSize)
            {
                std::cout << "\nOverflow: Could not insert\n";
                return;
            }

            heapSize++;
            int i = heapSize-1;
            arr[i] = x;

            while (i != 0 && arr[parentIndex(i)] < arr[i])
            {
                swap(i, parentIndex(i));
                i = parentIndex(i);
            }
        }

        int pop()
        {
            if (heapSize == 0) return INT_MIN;

            heapSize--;

            if (heapSize == 1) return arr[0];

            int root = arr[0];
            arr[0] = arr[heapSize];

            heapfy(0);
            return root;
        }

        void removeIndex(int index)
        {
            if (index >= heapSize || index < 0)
                std::cout << "\nIndex Out Of Inbound\n";

            increaseKey(index, INT_MAX);
            pop();
        }

        void print()
        {
            if (heapSize == 0)
            {
                std::cout << "\nEmpty Heap\n";
                return;
            }

            for(int i = 0; i < heapSize; i++)
            {
                std::cout << "|" << arr[i];
            }
            std::cout << "|\n";
        }
};

int main () {
    MaxHeap h(15);
    h.insert(3);
    h.insert(10);
    h.insert(12);
    h.insert(8);
    h.insert(2);
    h.insert(14);

    std::cout << "The current size of heap is " << h.size() << "\n";
    std::cout << "The current maximum element is " << h.getMax() << "\n";
    
    h.print();
    h.pop();
    h.print();

    h.insert(15);
    h.insert(5);
    std::cout << "The current size of heap is " << h.size() << "\n";
    std::cout << "The current maximum element is " << h.getMax() << "\n";
    h.print();
    h.removeIndex(4);
    h.print();
}